### PR TITLE
chore(oauth): migrate OAuth package to Typescript

### PR DESCRIPTION
**Pull request by gjaskiewicz**
# Summary

Migrate to OAuth package to Typescript

# Changes Made

chore(oauth): migrate to Typescript

# Related Issues

- [Issue 1:  #196 

# Testing Instructions

1. `npm install`
2. `turbo dev`

# Checklist

- [x] I have performed a self-review of my code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have made corresponding changes to the documentation (if applicable).
- [x] My changes generate no new warnings or errors.

### CODE REVIEW

File: packages/oauth/__tests__/mockedSession.test.ts [hunk 1]
```diff
@@ -53,9 +54,10 @@ describe("mockedSession", () => {
   })
 
   describe("session", () => {
-    let session, onUpdate
+    let session: any = undefined
```
Comment (barsukov) [inline packages/oauth/__tests__/mockedSession.test.ts:57]: we could also try to avoid it.

File: packages/oauth/package.json [hunk 2]
```diff
@@ -1,18 +1,17 @@
 {
   "name": "@cloudoperators/juno-oauth",
   "version": "1.2.7",
-  "description": "Authenticates user via OIDC",
   "author": "UI-Team",
-  "contributors": [
-    "Andreas Pfau"
-  ],
+  "description": "Authenticates user via OIDC",
+  "main": "build/oauth.es.js",
```
Comment (barsukov) [inline packages/oauth/package.json:6]: I think we need here to be and `build/index.js` everywhere

File: packages/oauth/package.json [hunk 3]
```diff
@@ -1,18 +1,17 @@
...
+  "module": "build/oauth.es.js",
```
Comment (barsukov) [inline packages/oauth/package.json:7]: I think we need here to be and build/index.js

File: packages/oauth/src/codeFlow.ts [hunk 4]
```diff
@@ -8,9 +8,9 @@ import { decodeIDToken } from "./tokenHelpers"
 import { searchParams } from "./oidcState"
 import { paramsToUrl } from "./utils"
 
-const exchangeCode = async ({ tokenEndpoint, code, verifier, clientID }) => {
+const exchangeCode = async ({ tokenEndpoint, code, verifier, clientID }: any) => {
   if (!clientID) throw new Error("clientID is required")
-  const body = {
+  const body: any = {
```
Comment (barsukov) [inline packages/oauth/src/codeFlow.ts:13]: lets try to fix it

File: packages/oauth/src/codeFlow.ts [hunk 5]
```diff
@@ -8,9 +8,9 @@ import { decodeIDToken } from "./tokenHelpers"
 import { searchParams } from "./oidcState"
 import { paramsToUrl } from "./utils"
 
-const exchangeCode = async ({ tokenEndpoint, code, verifier, clientID }) => {
+const exchangeCode = async ({ tokenEndpoint, code, verifier, clientID }: any) => {
```
Comment (barsukov) [inline packages/oauth/src/codeFlow.ts:11]: do you really need here any?

File: packages/oauth/src/oidcConfig.ts [hunk 6]
```diff
@@ -3,19 +3,25 @@
...
+interface CachedConfig {
+  time: number
+  config: Record<string, any>
+}
+
+let oidcConfig: Record<any, CachedConfig> = {}
+const cacheDuration = 5 * 60 * 60 * 1000
 
-export async function getOidcConfig(issuerURL) {
+export async function getOidcConfig(issuerURL: any): Promise<any> {
```
Comment (barsukov) [inline packages/oauth/src/oidcConfig.ts:14]: maybe we could try object.

File: packages/oauth/src/oidcSession.ts [hunk 7]
```diff
@@ -45,22 +45,22 @@ const createOidcRequest = async ({ issuerURL, clientID, flowType, requestParams
...
+      const params = typeof requestParams === "string" ? JSON.parse(requestParams) : requestParams
       const newUrl = new URL(url)
       Object.keys(params).forEach((k) => newUrl.searchParams.append(k, params[k]))
       url = newUrl.href
     }
 
     // redirect to this URL
     window.location.replace(url)
-  } catch (error) {
+  } catch (error: any) {
```
Comment (barsukov) [inline packages/oauth/src/oidcSession.ts:56]: lets create an error type

File: packages/oauth/__tests__/codeFlow.test.ts [hunk 8]
```diff
@@ -2,27 +2,30 @@
...
-    json: () => Promise.resolve({ id_token: testIdToken }),
-  })
-)
+const mockIdTokenResponse = {
+  ok: true,
+  statusText: "OK",
+  json: () => {
+    return { id_token: testIdToken }
+  },
+} as any
```
[the remark this answers is not in the document]
Comment (gjaskiewicz) [reply packages/oauth/__tests__/codeFlow.test.ts]: Done

File: packages/oauth/__tests__/mockedSession.test.ts [hunk 9]
```diff
@@ -105,9 +107,9 @@ describe("mockedSession", () => {
     })
 
     describe("custom token", () => {
-      let onUpdate
+      let onUpdate: any = undefined
```
[the remark this answers is not in the document]
Comment (gjaskiewicz) [reply packages/oauth/__tests__/mockedSession.test.ts]: Done, but has to make changes in tests

File: packages/oauth/src/codeFlow.ts (hunk 4 above)
Comment (gjaskiewicz) [reply packages/oauth/src/codeFlow.ts]: Done

File: packages/oauth/src/oidcSession.ts (hunk 7 above)
Comment (gjaskiewicz) [reply packages/oauth/src/oidcSession.ts]: Done, used unknown with type check

File: packages/oauth/eslint.config.mjs [hunk 10]
```diff
@@ -3,12 +3,27 @@
...
   {
-    files: ["__tests__/**"],
-    languageOptions: { sourceType: "module" },
+    files: ["**/*.ts", "**/*.tsx"],
+    languageOptions: {
+      parserOptions: {
+        project: ["./tsconfig.json"], // Ensure this points to your tsconfig.json
+      },
+    },
+    // TODO: We need to make all of this checks on again, step by step
```
[the remark this answers is not in the document]
Comment (gjaskiewicz) [reply packages/oauth/eslint.config.mjs:17]: no-unsafe-member-access - 110 errors - I leave this rule
no-floating-promises - 6 errors, added empty success and rejection handlers if missing
no-unsafe-argument - 30 errors, I leave this rule
no-unsafe-call - 38 errors, I leave this rule
no-unsafe-assignment - 142 errors, I leave this rule
require-await - 1 error, I added skip line exception

File: packages/oauth/__tests__/mockedSession.test.ts (hunk 1 above)
Comment (gjaskiewicz) [reply packages/oauth/__tests__/mockedSession.test.ts]: Done

File: packages/oauth/package.json (hunk 2 above)
Comment (barsukov) [reply packages/oauth/package.json:6]: Still need to be fixed :)

File: packages/oauth/package.json (hunk 3 above)
Comment (barsukov) [reply packages/oauth/package.json:7]: same here

File: packages/oauth/package.json (hunk 2 above)
Comment (gjaskiewicz) [reply packages/oauth/package.json]: Fixed in recent commit

File: packages/oauth/package.json (hunk 3 above)
Comment (gjaskiewicz) [reply packages/oauth/package.json]: Fixed

File: packages/oauth/src/codeFlow.ts (hunk 5 above)
Comment (gjaskiewicz) [reply packages/oauth/src/codeFlow.ts]: Changed to record

File: packages/oauth/src/oidcConfig.ts (hunk 6 above)
Comment (gjaskiewicz) [reply packages/oauth/src/oidcConfig.ts]: Added proper type

**Comment by barsukov** [comment]
@andypf would be nice if you could check it out the final version of sthis TS transition.

**Comment by barsukov** [comment]
Also please make a changeset :)

**Review by barsukov — APPROVED** [review]

**Pull request merged**